whelk-core/src/main/groovy/whelk/component/SparqlUpdater.groovy

```diff
@@ -0,0 +1,149 @@
+    private static final int READ_TIMEOUT_MS = 5 * 1000
+
+    // Number of items to take from queue each time
+    private static final int QUEUE_TAKE_NUM = 1
+
+    private final ExecutorService executorService
+    private final Runnable task
+    private final Timer timer = new Timer("${SparqlUpdater.class.getName()}-timer", true)
+
+    private final int numWorkers
+
+    static SparqlUpdater build(PostgreSQLComponent storage, Map jsonLdContext, Properties configuration) {
+        if (!(configuration.getProperty("sparqlEnabled")?.toLowerCase() == 'false')) {
+            storage.sparqlQueueEnabled = true
+        }
+
+        String sparqlCrudUrl = configuration.getProperty("sparqlCrudUrl")
+        if (sparqlCrudUrl) {
+            int numWorkers = configuration.getProperty("sparqlNumWorkers")
+                    ? Integer.parseInt(configuration.getProperty("sparqlNumWorkers"))
+                    : DEFAULT_NUM_WORKERS
+
+            Virtuoso virtuoso = new Virtuoso(
+                    jsonLdContext,
+                    buildHttpClient(numWorkers),
+                    sparqlCrudUrl,
+                    configuration.getProperty("sparqlUser"),
+                    configuration.getProperty("sparqlPass"))
+
+            return new SparqlUpdater(storage, virtuoso, numWorkers)
+        }
+        else {
+            return new SparqlUpdater()
+        }
+    }
+
+    private SparqlUpdater() {
+        executorService = null
+        task = null
+        numWorkers = 0
+    }
+
+    private SparqlUpdater(PostgreSQLComponent storage, Virtuoso sparql, int numWorkers) {
+        this.numWorkers = numWorkers
+        this.executorService = buildExecutorService(numWorkers)
+
+        PostgreSQLComponent.QueueHandler handler = { Document doc ->
+            try {
+                if (doc.deleted) {
+                    sparql.deleteNamedGraph(doc)
+                }
+                else {
+                    sparql.insertNamedGraph(doc)
+                }
+                return HANDLED
+            }
+            catch (UnexpectedHttpStatusException e) {
+                log.warn("Failed, will requeue: $e")
+                return FAIL_REQUEUE
+            }
+            catch (Exception e) {
+                log.warn("Failed, will retry: $e")
+                return FAIL_RETRY
+            }
+        }
+
+        DataSource connectionPool = storage.createAdditionalConnectionPool(this.getClass().getSimpleName(), numWorkers)
+        this.task = {
+            try {
+                // Run as long as there still might be docs in the queue and we haven't failed
+                if (storage.sparqlQueueTake(handler, QUEUE_TAKE_NUM, connectionPool)) {
+                    pollNow()
+                }
+            }
+            catch (Exception e) {
+                log.warn("Error executing task: $e", e)
+            }
+        }
+
+        timer.scheduleAtFixedRate({ pollNow() }, PERIODIC_CHECK_MS, PERIODIC_CHECK_MS)
```

Comment:
This calls into executorService.submit, which CAN throw, in which case (it seems) all future timer events are stopped for ever (until program restart).
pollNow() should probably have a try/catch{_everything_}.

Reply:
Good catch (pun intended)

Reply:
I think it should be safe since we're using `ThreadPoolExecutor.DiscardPolicy` but I will add a catch.